Change save path to /Library for Catalina or above (for System Integrity Protection) and add ability to restore system-wide settings

jnooree opened the pull request:
This commit will change destination directory from /System/Library to /Library for systems running macOS Catalina or above.
Also adds ability to handle system-provided settings more precisely, and fixes #15, a bug related to writing hiDPI flag(s).

2020/09/01 UPDATE:
Now this could automatically backup the system-wide settings, and has a menu to restore those.

### Major Changes
- Change save path from /System/Library to /Library for systems running macOS 10.15.0 or above
- Add ability to handle both /System/Library and /Library if (this feature is for backward-compatibility)
  - Current macOS version is 10.15.0 or above; and
  - System Integrity Protection is disabled
- Add ability to handle other 'hiDPI flags', which could be found in some system-provided plists in /System/Library
- Fix bug #15: now hiDPI flags are consistent with the macOS default values (also with comsysto/Display-Override-PropertyList-File-Parser-and-Generator-with-HiDPI-Support-For-Scaled-Resolutions@a7f0b42)

2020/09/01 UPDATE:
- Add ability to restore the system-wide settings under /System/Library (resolves #1)
- Add ability to select and remove multiple resolutions
- Bump version up to 2.4.0

2020/09/08 UPDATE:
- Introduce integrated aspect ratio calculator (works both way)
- More UI improvements
- Fix a bug that might cause app to crash, right after launch

### Minor Changes
- Split files for readability
- Change some function-specific variables into static variables, namely
  - defaultHiDPIFlag in Resolution
  - rootWriteable, afterCatalina, and dirformat in ViewController
- Add some methods to Resolution class

2020/09/01 UPDATE:
- Remove unused option in cmdline.mm
- Add some convenience functions

2020/09/08 UPDATE:
- Fix memory leak in RDMAppDelegate.mm

jnooree committed:
Change save path to /Library for Catalina or above

This commit will change destination directory from /System/Library
to /Library for systems running macOS Catalina or above.
Also adds ability to handle system-provided settings more precisely,
and fixes #15, a bug related to writing hiDPI flag(s).

Major Changes:
- Change save path from /System/Library to /Library for systems running
  macOS 10.15.0 or above
- Add ability to handle both /System/Library and /Library if
  * Current macOS version is 10.15.0 or above; and
  * System Integrity Protection is disabled
  (This feature is for backward-compatibility.)
- Add ability to handle other 'hiDPI values', which could be found in
  some system-provided plists in /System/Library
- Fix bug: now hiDPI flags are consistent with the macOS default
  values (also with comsysto/Display-Override-PropertyList-File-Parser-and-Generator-with-HiDPI-Support-For-Scaled-Resolutions@a7f0b42)

Minor Changes:
- Split files for readability
- Change some function-specific variables into static variables, namely
  * defaultHiDPIFlag in Resolution
  * rootWriteable, afterCatalina, and dirformat in ViewController
- Add some methods to Resolution class

jnooree committed:
Remove debug symbols and rename a variable

jnooree committed:
Edit xcode project settings

jnooree committed:
Fix possible errors related to temp filename

- Change temporary file name from tmp to Universal Unique ID (UUID)
- Improve UI
- Remove all deprecated functions

jnooree committed:
Cleanup and update xcode settings

jnooree committed:
Change readme

jnooree committed:
Change behavior when HiDPI flag is set

- Now, the resolution automatically doubled when HiDPI flag is set,
  so no maths anymore.
- Changes README.md to match the interface.
- Bumps version to 2.3.4 and remove trailing version info string.
- Moves Main.stroyboard to the project root.

jnooree committed:
Fix typo

jnooree committed:
Fix typo

jnooree committed:
Fix bugs and add install script

jnooree committed:
Update readme

jnooree committed:
Add newline

jnooree committed:
Add uninstall script

jnooree committed:
Cleanup

jnooree committed:
Update readme

jnooree committed:
Add ability to restore system-wide settings

Major Changes:
- Add ability to restore the system-wide settings under /System/Library
- Add ability to select and remove multiple resolutions
- Bump version up to 2.4.0

Minor Changes:
- Remove unused option in cmdline.mm
- Add some convenience functions

This commit resolves #1.

jnooree commented:
Also added ability to restore the system-wide settings.

### Major Changes
- Add ability to restore the system-wide settings under /System/Library
- Add ability to select and remove multiple resolutions
- Bump version up to 2.4.0

### Minor Changes 
- Remove unused option in cmdline.mm
- Add some convenience functions

This commit resolves #1.

jnooree committed:
Fix type matching

jnooree committed:
Fix bug in restoring settings

Major changes:
- Fix a bug related to restoring settings that could
  cause unexpected errors

Minor changes:
- Functions are more 'swifty' now
- Introduce faster decoding function
- Update uninstall success message

jnooree committed:
Improve performance

leonmz commented on src/EditDisplayPlistItem.swift:
```diff
@@ -9,14 +9,15 @@
 @objc class EditDisplayPlistItem : NSMenuItem {
-	@objc var vendorID : UInt32
-	@objc var productID : UInt32
+	@objc var vendorID	  : UInt32
```
Are these spaces intended?

jnooree replied:
Oh no... the spaces are tabs, and this would be aligned when translate tabs as four spaces. I'll change all internal blanks to spaces soon.

jnooree replied:
Fixed by commit jnooree/RDM@1cc0c78166aa879ee5a3d3be8bc32da69e6c4ecf

jnooree committed:
Improve performance and layouts

jnooree committed:
Rename SR- to RDM-

jnooree committed:
Introduce integrated aspect ratio calculator

Now RDM has integrated aspect ratio calculator. Calculate the ratio
in the calculator, and it would be added directly into the table view.

Major Changes:
- Introduce integrated aspect ratio calculator (works both way)
- More UI improvements
- Fix a bug that might cause app to crash, right after launch

Minor Changes:
- Fix memory leak in RDMAppDelegate.mm
- Add convenience functions

TODO: Add aspect ratio information in the resolution selection menu

jnooree committed:
Comment out debug symbols

jnooree committed:
Fix calculator behavior

usr-sse2 commented:
Thanks for such a great work! I'll test and review it next week and then merge.

usr-sse2 commented on src/utils.mm:
```diff
@@ -0,0 +1,87 @@
+
```
Files should be renamed using `git mv` to preserve modification history. Now it looks like the original file was deleted and you became the author of the whole file.

jnooree replied:
~~I'll try to rebase and separate commits to preserve history, then force-push. It would be ok?~~ Actually I've already did it (jnooree/RDM@f5f8ead8951cf7e30c2c394aa7caf7ebb0cd20d5). The problem is, [github does not track renamed files in commit log](https://github.community/t/show-file-history-for-renamed-files/1740/3). The file history still remains and you can verify it by `git log --follow src/any_file_to_check`.

usr-sse2 replied:
Ok, I see, GitHub displays it properly when looking on a single commit but fails for the joined diff of the whole PR.

usr-sse2 commented on src/ViewController.swift:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    @IBAction func calcAspectRatio(_ sender: Any) {
+        presentAsSheet(calcController)
```
The sheet should have a Cancel button. Currently it's possible to close it without adding a new resolution only using Esc key.

jnooree replied:
I'll add one soon. Thanks for feedback!

jnooree replied:
Resolved in jnooree/RDM@e99669b1727cc5fe6508a1655f46f4a6c78bb050.

usr-sse2 commented on src/ResMenuItem.h:
```diff
@@ -9,23 +9,25 @@ static inline CFDictionaryRef CGDisplayModeGetDictionary(CGDisplayModeRef mode)
+    
+    //CGDisplayModeRef _mode;
+    
+    int refreshRate;
```
MacBook Pro 16 [supports non-integer refresh rates](https://www.provideocoalition.com/apple-macbook-pro-16-supports-non-integer-framerates/).

jnooree replied:
Yes it does, but the retrieved data would always be an integer. Please refer to https://github.com/jnooree/RDM/blob/b6935468139b8a62c73ec822d2d5b60cc2cb47d9/src/utils.h#L32.

usr-sse2 replied:
Then it's interesting how it should work on whe MBP.

jnooree replied:
I'm on that mac, and it fails to display the correct refresh rate (eg. 59 Hz instead 60 Hz, etc.). I've tried several thing to make it work again, but unfortunately, none of them worked.

usr-sse2 commented on Assets.xcassets/Calculator.imageset/Contents.json:
```diff
@@ -0,0 +1,62 @@
+          "value" : "dark"
+        }
+      ],
+      "filename" : "Calculator-dark-1x.png",
```
[Template images](https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/custom-icons/) allow to use the same image file for both dark and light icon.
(Not necessary to fix, just for information)

jnooree replied:
I didn't knew this! Thanks very much.

jnooree replied:
Thanks for info. I'll change it to the template image.

jnooree replied:
Fixed in jnooree/RDM@e99669b1727cc5fe6508a1655f46f4a6c78bb050.

usr-sse2 replied:
You forgot to add the new image file.

jnooree replied:
Sorry 😅 Now it is really fixed in jnooree/RDM@e4102bc!

jnooree committed:
Replace calculator icon with template image

Also adds a cancel button to SheetViewController.

usr-sse2 commented:
Do systems older than Catalina support `/Library/Displays/Contents/Resources/Overrides`? If so, we should use that always instead of trying to write into the system folder. I just didn't know that `/Library/...` is supported.

jnooree committed:
Add missing image file

jnooree commented:
> Do systems older than Catalina support `/Library/Displays/Contents/Resources/Overrides`? If so, we should use that always instead of trying to write into the system folder. I just didn't know that `/Library/...` is supported.

I just have no idea. I only have a mac that supports Catalina and above, so I cannot verify it. I've found out that the `/Library/` folder works from [here](https://comsysto.github.io/Display-Override-PropertyList-File-Parser-and-Generator-with-HiDPI-Support-For-Scaled-Resolutions/), at least for Catalina and above.

jnooree committed:
Fix bugs in RestoreSettings.swift

Also improves the app support directory finding mechanism

jnooree committed:
Update project settings and bug fix in insall script

usr-sse2 commented:
Ok, then I'll check it myself on High Sierra.

usr-sse2 commented:
High Sierra ignores `/Library/Displays` (and at least in my case seems to ignore `scaled-resolutions`, so supporting it becomes useless in that case, but I'll check later with a non-TV display), so I'll check Mojave next.

usr-sse2 commented:
While testing I found that if `/System/Library/Displays/Contents/Resources/Overrides/DisplayVendorID-30ae/DisplayProductID-60c9` doesn't exist, then RDM fails because it can't backup this file. Only when I created this file, RDM allowed to save settings in `/Library/Displays/Contents/Resources/Overrides/DisplayVendorID-30ae/DisplayProductID-60c9`.

jnooree committed:
Fix bug related to backup

jnooree commented:
> While testing I found that if `/System/Library/Displays/Contents/Resources/Overrides/DisplayVendorID-30ae/DisplayProductID-60c9` doesn't exist, then RDM fails because it can't backup this file. Only when I created this file, RDM allowed to save settings in `/Library/Displays/Contents/Resources/Overrides/DisplayVendorID-30ae/DisplayProductID-60c9`.

It is now fixed by jnooree/RDM@9b8623b. Sorry for my mistake.

jnooree commented:
One small thing to note: the version `2.4.0` should be added to the latest commit (or the commit that has changed the internal version string, c61cbe4), to match the documentation. Thanks for merging my PR!

usr-sse2 commented:
I'm not releasing it now, so I'll update the tags and docs later. There are issues with the calculator and with working with both plist locations (we shouldn't write into System folder on 10.15+ even if it's writable) that I'll fix myself.
